### ISSUE TITLE
Write performance seems terrible? Am I holding it wrong?

### ISSUE THREAD

**Issue by SirVer** [spacejam/sled#334]
I am developing a large scale point cloud viewer. It uses an on-disk Octree which in the current design is spread in thousand of files in a directory. I want to move towards an embedded database for storage and investigated a few options. I'd like to use sled, but I run into a terrible performance problem.

I wrote a tiny tool that just takes files from a directory and puts them into a database, i.e. convert from the current representation into the one that I'd like to use. The `key` is the basename of the files (up to 20 chars of len) and the `value` is the content of the file (up to one MB) in my test case. There are 15368 files in that directory. 

I wrote this [tool once for sled](https://github.com/SirVer/point_cloud_viewer/blob/00_sled/src/bin/to_rocksdb.rs) and [once for lmdb](https://github.com/SirVer/point_cloud_viewer/blob/00_lmdb/src/bin/to_rocksdb.rs). Running the one for lmdb takes ~30 seconds, the sled one gets ever slower and the ETA is days. 

Could you look at the 50 lines of code and tell me if I am doing something very stupid? Or is my use case unreasonable for sled?

--------------
1. sled version: "0.15.17"
2. rustc version:
~~~
$ cargo --version
cargo 1.29.0-nightly (b42488270 2018-08-04)
~~~

3. operating system: Mac OS 10.13.4.
4. logs: Not sure what this is here.

**Comment by spacejam** [comment]
Sled has an architecture that will eventually shatter the write performance of lmdb due to supporting writes concurrently from several threads. 

However, sled has received very little performance tuning up until now. There is so much low-hanging fruit to speed up the write path. For large values that you're talking about, sled is making several copies of the data in the write path. Large values exacerbates this problem significantly. When a calling thread triggers an IO buffer to be flushed, it blocks as it waits for this to complete. 

If you send me a flamegraph I can likely optimize the biggest bottleneck that your code is running into. Check out the `hack/flamerun.sh` script in this repo which will generate one, you can copy the `hack` directory to whatever directory you run your program from and type `hack/flamerun.sh myprog`

**Comment by SirVer** [comment]
That script does not seem to work on Mac OS since it requires `perf`, which does not seem to exist for Mac. I will try on a Linux box when I have access to one.

**Comment by ehiggs** [comment]
AIUI, FlameGraph is based on @brendangregg's work here: https://github.com/brendangregg/FlameGraph

The `hack/flamegraph.sh` script looks like it's copypasta directly from the README so maybe make a new `hack/flamerun.macos.sh` with the following:

```bash
#!/bin/bash
set -ex
./"${1}" &
PID=$!
dtrace -x ustackframes=100 -n "profile-97 /pid == $PID/ { @[ustack()] = count(); } tick-60s { exit(0); }" -o out.user_stacks
wait "$PID"
./hack/stackcollapse.pl out.user_stacks > out.folded
./hack/flamegraph.pl out.folded > flamegraph.svg
rm out.user_stacks out.folded
```

You'll also need stackcollapse.pl from the flamegraph repo.

Alas when I test this, I only get addresses listed and not the actual function names.

**Comment by spacejam** [comment]
@SirVer there have been a few improvements to how writes are handled, but I'm not sure if it would address your issue yet. I've added an OSX-compatible flamegraph generator in `hack/flamerun_osx.sh` and it can be run by running this command from the root of the sled repo:
```
./hack/flamerun_osx.sh /path/to/your/slow/program
```

I'd love to look at the `flamegraph.svg` that it generates for your workload with a current sled release (0.16 is out now). If you have time to run this, you can either email me the svg that it prints out at t@jujit.su or just attach a tar / zip to a github comment (I wish you could attach svg without this hoop).

**Comment by SirVer** [comment]
@spacejam Sorry for not getting back to this earlier. But since sled does not seem to be ready for our uses yet, I had to bump this down in priority.

These are my experiments done today: I updated sled to 16.2 in my Cargo.toml at which point I could no longer compile my project:

~~~
cargo build --release
    Updating crates.io index
   Compiling sled_sync v0.1.2
   Compiling pagecache v0.7.3
   Compiling sled v0.16.2
error[E0412]: cannot find type `Guard` in this scope
  --> /Users/hrapp/.cargo/registry/src/github.com-1ecc6299db9ec823/sled-0.16.2/src/pinned_value.rs:11:42
   |
11 | pub struct PinnedValue(*const u8, usize, Guard);
   |                                          ^^^^^ not found in this scope

error[E0412]: cannot find type `Guard` in this scope
  --> /Users/hrapp/.cargo/registry/src/github.com-1ecc6299db9ec823/sled-0.16.2/src/pinned_value.rs:46:40
   |
46 |     pub(crate) fn new(v: &[u8], guard: Guard) -> PinnedValue {
   |                                        ^^^^^ not found in this scope

error[E0603]: struct `Guard` is private
 --> /Users/hrapp/.cargo/registry/src/github.com-1ecc6299db9ec823/sled-0.16.2/src/iter.rs:3:17
  |
3 | use pagecache::{Guard, Measure, M};
  |                 ^^^^^

error[E0603]: struct `Guard` is private
 --> /Users/hrapp/.cargo/registry/src/github.com-1ecc6299db9ec823/sled-0.16.2/src/tree.rs:6:22
  |
6 | use pagecache::{pin, Guard, PagePtr};
  |                      ^^^^^

error[E0603]: function `debug_delay` is private
   --> /Users/hrapp/.cargo/registry/src/github.com-1ecc6299db9ec823/sled-0.16.2/src/tree.rs:782:9
    |
782 |         pagecache::debug_delay();
    |         ^^^^^^^^^^^^^^^^^^^^^^

error[E0603]: function `debug_delay` is private
   --> /Users/hrapp/.cargo/registry/src/github.com-1ecc6299db9ec823/sled-0.16.2/src/tree.rs:795:9
    |
795 |         pagecache::debug_delay();
    |         ^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 6 previous errors

Some errors occurred: E0412, E0603.
For more information about an error, try `rustc --explain E0412`.
error: Could not compile `sled`.

To learn more, run the command again with --verbose.
~~~

This error went away after switching to `sled = { git = "https://github.com/spacejam/sled", rev = "2c146a7a" }`. This require a few trivial adaptations in my code since the API of sled has changed. They are all in the same branch then linked above.

After this the performance seemed much better: It still slowed down tremendously, starting with an estimate of 5min, but after total progress of 10% it was already at an estimate of 1h. 

I then ran the hack tool:

~~~
../sled/hack/flamerun.sh target/release/to_rocksdb
++ dirname ../sled/hack/flamerun.sh
+ DIR=../sled/hack
+ case `uname -s` in
++ uname -s
+ proc_name=to_rocksdb
+ echo 'expecting process name to be to_rocksdb'
expecting process name to be to_rocksdb
+ sudo dtrace -c target/release/to_rocksdb -o out.stacks -n 'profile-997 /execname == "to_rocksdb"/ { @[ustack(100)] = count(); }'
Password:
dtrace: system integrity protection is on, some features will not be available

dtrace: description 'profile-997 ' matched 1 probe
~~~

I let it run for a few minutes, until it chewed through 10% of my input data and estimated an ETA of 1h. The flamegraph is attached.
[flamegraph.svg.zip](https://github.com/spacejam/sled/files/2567442/flamegraph.svg.zip)

**Comment by spacejam** [comment]
@SirVer thank you so much for the flamegraph! That's really helpful for getting this taken care of :] I'm pretty sure this is the same issue as in #402, which will be resolved after some correctness issues are ironed out.

Just pushed pagecache `0.7.4` which irons out the visibility issue you just hit, so you can avoid the workaround by just removing your `Cargo.lock` or bumping up pagecache. Lots of dependencies are being moved around as all of the concurrency primitives are being put into a runtime module that will allow different kinds of executors to run the system in stressful ways.

Yes, I mean "alpha" when I say it :P But explorations like yours help to beat the thing into a good place over time, so I'm quite grateful for the report!

I'll circle back with you when I have fixed #402!

**Comment by SirVer** [comment]
> Yes, I mean "alpha" when I say it :P But explorations like yours help to beat the thing into a good place over time, so I'm quite grateful for the report!

Yes, do not worry. I do not expect anything else when reading alpha and I saw these experiments I ran as likely not leading to a production solution. I am more than happy to provide data to improve open source projects! I have been in your shows often enough and know how important early adopters feedback is :). Just the prioritization means my cycle times will be longer.

> I'll circle back with you when I have fixed #402!

👍
